kore/src/Kore/Internal/Predicate.hs

```diff
@@ -119,438 +184,642 @@ import Pretty
+    :: InternalVariable variable
+    => Predicate variable
+    -> Predicate variable
+    -> (Predicate variable, HasChanged)
+makeOrPredicate' p1 p2
+  | isTop p1 = (p1, Changed)
+  | isTop p2 = (p2, Changed)
+  | isBottom p1 = (p2, Changed)
+  | isBottom p2 = (p1, Changed)
+  | p1 == p2 = (p1, Changed)
+  | otherwise = (synthesize $ OrF Or
+    { orSort = ()
+    , orFirst = p1
+    , orSecond = p2
+    }
+    , NotChanged)
 
-{-| 'makeOrPredicate' combines two Predicates with an 'or', doing
-some simplification.
--}
 makeOrPredicate
     :: InternalVariable variable
     => Predicate variable
     -> Predicate variable
     -> Predicate variable
-makeOrPredicate t@PredicateTrue _ = t
-makeOrPredicate _ t@PredicateTrue = t
-makeOrPredicate PredicateFalse second = second
-makeOrPredicate first PredicateFalse = first
-makeOrPredicate first second
-  | first == second = first
-  | otherwise =
-    binaryOperator TermLike.mkOr first second
+makeOrPredicate p1 p2 = fst (makeOrPredicate' p1 p2)
+
+makeImpliesPredicate'
+    :: InternalVariable variable
+    => Predicate variable
+    -> Predicate variable
+    -> (Predicate variable, HasChanged)
+makeImpliesPredicate' p1 p2
+  | isBottom p1 = (makeTruePredicate, Changed)
+  | isTop p2 = (p2, Changed)
+  | isTop p1 = (p2, Changed)
+  | isBottom p2 = (makeNotPredicate p1, Changed)
+--  | p1 == p2 = (makeTruePredicate, Changed)
+  | otherwise = (synthesize $ ImpliesF Implies
+    { impliesSort = ()
+    , impliesFirst = p1
+    , impliesSecond = p2
+    }
+    , NotChanged)
 
-{-| 'makeImpliesPredicate' combines two Predicates into an
-implication, doing some simplification.
--}
 makeImpliesPredicate
     :: InternalVariable variable
     => Predicate variable
     -> Predicate variable
     -> Predicate variable
-makeImpliesPredicate PredicateFalse _ = makeTruePredicate_
-makeImpliesPredicate _ t@PredicateTrue = t
-makeImpliesPredicate PredicateTrue second = second
-makeImpliesPredicate first PredicateFalse = makeNotPredicate first
-makeImpliesPredicate first second =
-    binaryOperator TermLike.mkImplies first second
+makeImpliesPredicate p1 p2 = fst (makeImpliesPredicate' p1 p2)
 
-{-| 'makeIffPredicate' combines two evaluated with an 'iff', doing
-some simplification.
--}
-makeIffPredicate
+makeIffPredicate'
     :: InternalVariable variable
     => Predicate variable
     -> Predicate variable
-    -> Predicate variable
-makeIffPredicate PredicateFalse second = makeNotPredicate second
-makeIffPredicate PredicateTrue second = second
-makeIffPredicate first PredicateFalse = makeNotPredicate first
-makeIffPredicate first PredicateTrue = first
-makeIffPredicate first second =
-    binaryOperator TermLike.mkIff first second
+    -> (Predicate variable, HasChanged)
+makeIffPredicate' p1 p2
+  | isBottom p1 = (makeNotPredicate p2, Changed)
+  | isTop p1 = (p2, Changed)
+  | isBottom p2 = (makeNotPredicate p1, Changed)
+  | isTop p2 = (p1, Changed)
+--  | p1 == p2 = (makeTruePredicate, Changed)
```

Comment:
Why is this commented out?

Reply:
For two reasons:
1. It might be inefficient to compare two entire predicate trees when calling this maker function (which I believe happens quite often).
2. This is not a simplification that existed in the original code, so this would change the behavior of the function. It might not break anything but it requires testing.

I left it in as a comment in case we decide at some point that we do want it.

Reply:
Please remove the comment.